compiler/src/codegen/compcore.ml

```diff
@@ -1268,11 +1732,18 @@ let rec compile_store env binds =
   let process_binds env =
     let process_bind (b, instr) acc =
       let store_bind = compile_bind ~action:BindSet env b in
+      let store_bind_no_incref = compile_bind ~action:BindSet ~skip_incref:true env b in
       let get_bind = compile_bind ~action:BindGet env b in
-      let compiled_instr = match instr with
+      let compiled_instr, store_bind = match instr with
         | MAllocate(MClosure(cdata)) ->
-          allocate_closure env ~lambda:get_bind cdata
-        | _ -> compile_instr env instr in
+          (allocate_closure env ~lambda:get_bind cdata), store_bind_no_incref
+        (* HACK: We expect values returned from functions to have a refcount of 1, so we don't increment it when storing *)
+        (* | (MCallIndirect _) *)
+        | (MCallKnown _)
+        | (MAllocate _) -> (compile_instr env instr), store_bind_no_incref
+        (* [TODO] I think this is wrong? See commented out line above *)
+        | (MCallIndirect _)
```

Comment:
I think this is a TODO that we'd want to figure out now rather than merge it.

Reply:
As discussed, this is now noted under #152 .